qutip/solver/nonmarkov/transfertensor.py

```diff
@@ -0,0 +1,171 @@
+# -*- coding: utf-8 -*-
+# @author: Arne L. Grimsmo
+# @email1: arne.grimsmo@gmail.com
+# @organization: University of Sherbrooke
+
+"""
+This module contains an implementation of the non-Markovian transfer tensor
+method (TTM), introduced in [1].
+
+[1] Javier Cerrillo and Jianshu Cao, Phys. Rev. Lett 112, 110401 (2014)
+"""
+
+import numpy as np
+import time
+from qutip import spre, vector_to_operator, operator_to_vector, Result
+
+
+def ttmsolve(dynmaps, state0, times, e_ops=[], options=None):
+    """
+    Expand time-evolution using the Transfer Tensor Method [1]_, based on a set
+    of precomputed dynamical maps.
+
+    Parameters
+    ----------
+    dynmaps : list of :class:`qutip.Qobj`, callable
+        List of precomputed dynamical maps (superoperators) for the first times
+        of ``times`` or a callback function that returns the superoperator at a
+        given time.
+
+    state0 : :class:`qutip.Qobj`
+        Initial density matrix or state vector (ket).
+
+    times : array_like
+        List of times :math:`t_n` at which to compute results.
+        Must be uniformily spaced.
+
+    e_ops : :class:`qutip.qobj`, callable, or list.
+        Single operator or list of operators for which to evaluate
+        expectation values or callable or list of callable.
+        Callable signature must be, `f(t: float, state: Qobj)`.
+        See :func:`expect` for more detail of operator expectation.
+
+    options : dictionary
+        Dictionary of options for the solver.
+
+        - store_final_state : bool
+          Whether or not to store the final state of the evolution in the
+          result class.
+        - store_states : bool, None
+          Whether or not to store the state vectors or density matrices.
+          On `None` the states will be saved if no expectation operators are
+          given.
+        - normalize_output : bool
+          Normalize output state to hide ODE numerical errors.
+        - threshold : float
+          Threshold for halting. Halts if  :math:`||T_{n}-T_{n-1}||` is below
+          treshold.
+        - num_learning : int
+          Number of times used to construct the dynmaps operators when
+          ``dynmaps`` is a callable.
```

Comment:
Maybe this shouldn't be an option since it's not optional if `dynmaps` is callable?